### PR TITLE
NPM package of the Topologymodeler

### PR DESCRIPTION
**Pull request by ledicjp**
# NPM package @winery/topologymodeler

This adds the possibility to package the topologymodeler as npm package with the use of [ngpackagr](https://www.npmjs.com/package/ng-packagr). This change will make it possible to use the topologymodeler in different contexts, i.e, by passing the data to it directly or by configuring it to fetch the data from an endpoint by itself. Also updates Angular version to 6 so it becomes easier to use this as dependency in latest angular-cli generated projects, which already use Angular 6.

- Start and end date: 2018-05-01 - 2018-07-01

- Contributors: Josip Ledic, @ledicjp

- Supervisor: Karoline Saatkamp, @saatkamp; Lukas Harzenetter, @lharzenetter;  Oliver Kopp, @koppor 

## Current state of this:

- [x] new branch [topologymodeler-packaging](https://github.com/opentosca/winery/tree/topologymodeler-packaging)

- [x] Successfully packaged with ng-packagr ✅ 

- [x] Created new npmjs organization "Winery" to publish this package on: [@winery/topologymodeler](https://www.npmjs.com/package/@winery/topologymodeler)

- [x] Make endpoints configurable (see npm link for documentation)

- [x] Possibility to configure a "readonly" variant (see npm link for documentation)

- [x] Possibility to configure an "auto-layout on startup" option

- [x] Enhance & clean up bundling process (exclude unnecessary code from package, etc.)

- [x] Fully compatible with Angular 6

- [x] Added hotkeys and hotkey cheatsheet

- [x] refined styling of multiple components

- [x] added some more elements that improve the UX
    - e.g., popovers instead of directly opening a new tab
    - graphical delete buttons for relationship templates and node templates
    - fixing the issue where clicking the label of a relationship template wouldn't result in a selection, etc...

## Missing steps:

- [ ] CI solution which automatically bundles the latest build that passes and updates the version on npm

- [x] Ensure that Travis CI passes

- [ ] Ensure that you followed http://eclipse.github.io/winery/dev/ToolChain#github---prepare-pull-request.

- [ ] Branch name checked. That means, the branch name starts with `thesis/`, `fix/`, ...

- [x] Ensure that the commit message is [a good commit message](https://github.com/joelparkerhenderson/git_commit_message)

- [x] Ensure to use auto format in **all** files

- [ ] Change in CHANGELOG.md described

### CODE REVIEW

**Comment by genie-winery** [comment]
Can one of the admins verify this patch?

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/angular.json [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            }
+          }
+        },
+        "extract-i18n": {
+          "builder": "@angular-devkit/build-angular:extract-i18n",
+          "options": {
+            "browserTarget": "tm-dependency-test-app:build"
+          }
+        },
+        "test": {
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/angular.json]: You can remove this for now.

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/e2e/app.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { AppPage } from './app.po';
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/e2e/app.e2e-spec.ts]: You can remove this file for now.

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/e2e/app.po.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { browser, by, element } from 'protractor';
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/e2e/app.po.ts]: You can remove this file for now.

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/e2e/tsconfig.e2e.json [hunk 4]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/e2e/tsconfig.e2e.json]: You can remove this file for now.

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/package.json [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    "@angular/platform-browser-dynamic": "6.0.4",
+    "@angular/router": "6.0.4",
+    "angular2-hotkeys": "2.0.4",
+    "bootstrap": "4.1.1",
+    "core-js": "2.5.6",
+    "css-element-queries": "0.4.0",
+    "element-resize-detector": "1.1.12",
+    "elkjs": "0.0.1",
+    "font-awesome": "4.7.0",
+    "jsplumb": "2.6.12",
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/package.json]: Please update to `2.7.9` to be consistent with the workflowmodeler

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/package.json [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    "start": "ng serve",
+    "build": "ng build --prod",
+    "test": "ng test",
+    "lint": "ng lint",
+    "e2e": "ng e2e"
+  },
+  "private": true,
+  "dependencies": {
+    "@angular-redux/store": "6.5.7",
+    "@angular/animations": "6.0.4",
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/package.json]: In general: please ensure that you use the same versions as in the repository ui and the workflowmodeler.

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/app/app.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import { TestBed, async } from '@angular/core/testing';
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/app/app.component.spec.ts]: You can remove this file for now.

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/test.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+// This file is required by karma.conf.js and loads recursively all the .spec and framework files
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/test.ts]: You can remove this file for now.

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/tsconfig.spec.json [hunk 9]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/tsconfig.spec.json]: You can remove this file for now.

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/ttopology-template.ts [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+export class Visuals {
+
+    constructor(public color: string,
+                public nodeTypeId: string,
+                public localName?: string,
+                public imageUrl?: string) {
+    }
+
+    /*
+     get color(): string {
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/ttopology-template.ts]: If these methods are not used, remove them.

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/tslint.json [hunk 11]
```diff
@@ -0,0 +1,142 @@
+{
+  "rulesDirectory": [
+    "node_modules/codelyzer"
+  ],
+  "rules": {
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/tslint.json]: Similar to the original app, you can reuse the root tslint file.

File: org.eclipse.winery.topologymodeler.ui/angular.json [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+                        }
+                    }
+                },
+                "extract-i18n": {
+                    "builder": "@angular-devkit/build-angular:extract-i18n",
+                    "options": {
+                        "browserTarget": "org.eclipse.winery.topologymodeler.ui:build"
+                    }
+                },
+                "test": {
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/angular.json]: You can remove this for now.

File: org.eclipse.winery.topologymodeler.ui/angular.json [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+                        "tsConfig": [
+                            "src/tsconfig.app.json",
+                            "src/tsconfig.spec.json"
+                        ],
+                        "exclude": []
+                    }
+                }
+            }
+        },
+        "org.eclipse.winery.topologymodeler.ui-e2e": {
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/angular.json]: You can remove this for now.

File: org.eclipse.winery.topologymodeler.ui/lib/package.json [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        "lib-readme-mac": "cp ../src/doc/lib/README.md dist",
+        "packagr-mac": "ng-packagr -p ng-package.json && npm run lib-readme-mac",
+        "packagr": "ng-packagr -p ng-package.json && npm run lib-readme"
+    },
+    "private": false,
+    "peerDependencies": {
+        "@angular-redux/store": "6.5.7",
+        "@angular/animations": "6.0.4",
+        "@angular/cdk": "6.2.1",
+        "@angular/common": "6.0.4",
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/lib/package.json]: Similar as mentioned above - please ensure that you use the same versions as in the other projects.

File: org.eclipse.winery.topologymodeler.ui/lib/tslint.json [hunk 15]
```diff
@@ -0,0 +1,130 @@
+{
+  "rulesDirectory": [
+    "node_modules/codelyzer"
+  ],
+  "rules": {
+    "arrow-return-shorthand": true,
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/lib/tslint.json]: Does the extension of the root tslint file not work?

File: org.eclipse.winery.topologymodeler.ui/package.json [hunk 16]
```diff
@@ -1,6 +1,6 @@
 {
-    "name": "org.eclipse.winery.topologymodeler.ui",
-    "version": "2.0.0-SNAPSHOT",
+    "name": "topologymodeler",
+    "version": "2.1.0",
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/package.json]: MUST be the same as the root project - please revert the changes.

File: org.eclipse.winery.topologymodeler.ui/package.json [hunk 17]
```diff
@@ -1,6 +1,6 @@
 {
-    "name": "org.eclipse.winery.topologymodeler.ui",
-    "version": "2.0.0-SNAPSHOT",
+    "name": "topologymodeler",
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/package.json]: I would use `winery-topologymodeler`

File: org.eclipse.winery.topologymodeler.ui/package.json [hunk 18]
```diff
@@ -11,51 +11,79 @@
...
-        "@angular/core": "4.4.7",
-        "@angular/forms": "4.4.7",
-        "@angular/http": "4.4.7",
-        "@angular/platform-browser": "4.4.7",
-        "@angular/platform-browser-dynamic": "4.4.7",
-        "@angular/router": "4.4.7",
+        "@angular/animations": "6.0.4",
+        "@angular/cdk": "6.2.1",
+        "@angular/common": "6.0.4",
+        "@angular/core": "6.0.4",
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/package.json]: See above

File: org.eclipse.winery.topologymodeler.ui/src/app/canvas/entities-modal/entities-modal.service.ts [hunk 19]
```diff
@@ -12,12 +12,14 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
  ********************************************************************************/
 
+
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/src/app/canvas/entities-modal/entities-modal.service.ts]: Please delete the extra line - should be fixed by reformatting all code

File: org.eclipse.winery.topologymodeler.ui/src/app/canvas/entities-modal/entities-modal.service.ts [hunk 20]
```diff
@@ -44,8 +46,8 @@ export class EntitiesModalService {
         } else {
             URL = backendBaseURL + '/admin/namespaces/';
         }
-        return this.http.get(URL, this.options)
-            .map(res => res.json());
+        return this.http.get(URL, this.options).pipe(
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/src/app/canvas/entities-modal/entities-modal.service.ts]: Please refactor the `Http` to the `HttpClient`

File: org.eclipse.winery.topologymodeler.ui/src/app/node/node.component.ts [hunk 21]
```diff
@@ -284,7 +280,6 @@ export class NodeComponent implements OnInit, AfterViewInit, OnDestroy, DoCheck
 
     /**
      * Checks if it was a click or a drag operation on the node.
-     * @param $event
      */
     mouseUpHandler($event): void {
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/src/app/node/node.component.ts]: Please assign the `$event` its correspnding type. Applies for the whole file.

File: org.eclipse.winery.topologymodeler.ui/src/app/palette/palette.component.ts [hunk 22]
```diff
@@ -107,7 +107,13 @@ export class PaletteComponent implements OnInit, OnDestroy {
...
      */
     ngOnInit() {
-        // console.log(this.entityTypes);
+    }
+
+    /**
+     * Angular lifecycle event.
+     */
+    ngAfterViewInit() {
+        console.log(this.entityTypes);
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/src/app/palette/palette.component.ts]: Please remove this method, if it isn't essential.

File: org.eclipse.winery.topologymodeler.ui/src/app/palette/palette.component.ts [hunk 23]
```diff
@@ -236,19 +242,23 @@ export class PaletteComponent implements OnInit, OnDestroy {
      * @param name
      * @return result
      */
-    private getNewNodeDataFromNodeTypes(name: string): any {
+    public getNewNodeDataFromNodeTypes(name: string): any {
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/src/app/palette/palette.component.ts]: Don't you know the type of the return?

File: org.eclipse.winery.topologymodeler.ui/src/app/services/jsPlumb.service.ts [hunk 24]
```diff
@@ -14,34 +14,32 @@
...
-                Endpoint: 'Blank',
-                connectorOverlays: [
-                    ['Arrow', { location: 1 }],
+                Endpoint: "Blank",
+                ConnectionOverlays: [
+                    ["Arrow", { location: 1 }],
                 ],
                 ConnectionsDetachable: false,
-                Anchor: 'Continuous'
+                Anchor: "Continuous"
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/src/app/services/jsPlumb.service.ts]: Please use single quotes.

File: org.eclipse.winery.topologymodeler.ui/src/doc/lib/README.md [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+    readonly id: string;
+    readonly ns: string;
+    readonly repositoryURL: string;
+    readonly uiURL: string;
+    readonly compareTo: string;
+}
+```
+
+### Attention: this package has the following `peerDependencies`:
+##### Please make sure that you install them manually or by adding them to your `dependencies`.
```
Comment (lharzenetter) [inline org.eclipse.winery.topologymodeler.ui/src/doc/lib/README.md]: This is not a heading. Just print it in bold using `**text**`.

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/angular.json (hunk 1 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/angular.json]: done

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/e2e/app.e2e-spec.ts (hunk 2 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/e2e/app.e2e-spec.ts]: done

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/e2e/app.po.ts (hunk 3 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/e2e/app.po.ts]: done

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/e2e/tsconfig.e2e.json (hunk 4 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/e2e/tsconfig.e2e.json]: done

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/package.json (hunk 5 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/package.json]: done

File: org.eclipse.winery.topologymodeler.ui/src/app/node/node.component.ts (hunk 21 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/src/app/node/node.component.ts]: Types are invoked automatically. Adding type annotations manually causes the typescript compiler to abort during bundling.

File: org.eclipse.winery.topologymodeler.ui/src/app/palette/palette.component.ts (hunk 22 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/src/app/palette/palette.component.ts]: done

File: org.eclipse.winery.topologymodeler.ui/src/app/palette/palette.component.ts (hunk 23 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/src/app/palette/palette.component.ts]: This object is used only in this method and has no Class/Type right now

File: org.eclipse.winery.topologymodeler.ui/src/doc/lib/README.md (hunk 25 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/src/doc/lib/README.md]: done

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/test.ts (hunk 8 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/test.ts]: done

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/tsconfig.spec.json (hunk 9 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/tsconfig.spec.json]: done

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/ttopology-template.ts (hunk 10 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/src/ttopology-template.ts]: They are used depending on what you want to test, i.e., mock data vs. fetch data

File: org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/tslint.json (hunk 11 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/TmDependencyTestApp/tslint.json]: done

File: org.eclipse.winery.topologymodeler.ui/lib/package.json (hunk 14 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/lib/package.json]: done

File: org.eclipse.winery.topologymodeler.ui/lib/tslint.json (hunk 15 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/lib/tslint.json]: This tslint file is very likely never used

File: org.eclipse.winery.topologymodeler.ui/package.json (hunk 16 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/package.json]: done

File: org.eclipse.winery.topologymodeler.ui/package.json (hunk 17 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/package.json]: I will leave the original name "org.eclipse.winery.topologymodeler.ui" so it is consistent with the ui and workflowmodeler names

File: org.eclipse.winery.topologymodeler.ui/package.json (hunk 18 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/package.json]: done

File: org.eclipse.winery.topologymodeler.ui/src/app/canvas/entities-modal/entities-modal.service.ts (hunk 20 above)
Comment (ledicjp) [reply org.eclipse.winery.topologymodeler.ui/src/app/canvas/entities-modal/entities-modal.service.ts]: 😭

**Comment by lharzenetter** [comment]
CQ created as [CQ 16867](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=16867)
